File: cpp/091-100/Interleaving String.cpp
class Solution {
    
public:
    bool isInterleave(string s1, string s2, string s3) {
        if(s1.length() + s2.length() != s3.length())
            return false;
        
        bool table[s1.length()+1][s2.length()+1];
        for(int i = 0;i <= s1.length();i++)
            for(int j = 0;j <= s2.length();j++)
                table[i][j] = false;

        
        table[0][0] = true;
        for(int i = 1;i <= s1.length();i++)
            if(s3[i-1] == s1[i-1] && table[i-1][0])
                table[i][0] = true;

        for(int j = 1;j <= s2.length();j++)
            if(s3[j-1] == s2[j-1] && table[0][j-1])
                table[0][j] = true;
        
        for(int i = 1;i <= s1.length();i++)
        {
            for(int j = 1;j <= s2.length();j++)
            {
                if(s3[i+j-1] == s1[i-1] && table[i-1][j] == true)
                    table[i][j] = true;
                if(s3[i+j-1] == s2[j-1] && table[i][j-1] == true)
                    table[i][j] = true;
            }
        }

        return table[s1.length()][s2.length()];
    }
    
};

File: python/091-100/Interleaving String.py
class Solution:
    # @param {string} s1
    # @param {string} s2
    # @param {string} s3
    # @return {boolean}
    def isInterleave(self, s1, s2, s3):
        m = len(s1)
        n = len(s2)
        if m+n != len(s3):
            return False
        
        table = [([False] * (m+1)) for i in range(n+1)]
        table[0][0] = True
        for i in range (1, m+1):
            if s3[i-1] == s1[i-1] and table[0][i-1] == True:
                table[0][i] = True
        for i in range (1, n+1):
            if s3[i-1] == s2[i-1] and table[i-1][0] == True:
                table[i][0] = True
    
        for i in range (1, n+1):
            for j in range(1, m+1):
                if s3[i+j-1] == s2[i-1] and table[i-1][j] == True:
                    table[i][j] = True
                if s3[i+j-1] == s1[j-1] and table[i][j-1] == True:
                    table[i][j] = True

                        return table[n][m]